replica/table: add tablet count metric

pwrobelse opened the pull request:
This change introduces a new metric called tablet_count
that is recalculated during construction of table object
and on each call to table::update_effective_replication_map().

To get the count of tablet per current shard, tablet map
is traversed and for each tablet_id tablet_map::get_shard()
is called. Its return value is compared with this_shard_id().

The new metric is maintained and exposed only for tables
that uses tablets.

Sample output from run with  `--enable-keyspace-column-family-metrics true`:
```
# HELP scylla_column_family_tablet_count Tablet count
# TYPE scylla_column_family_tablet_count gauge
scylla_column_family_tablet_count{cf="mytable",ks="testing",shard="0"} 1.000000
scylla_column_family_tablet_count{cf="mytable2",ks="testing",shard="0"} 1.000000
scylla_column_family_tablet_count{cf="mytable3",ks="testing",shard="0"} 1.000000
scylla_column_family_tablet_count{cf="mytable",ks="testing",shard="1"} 1.000000
scylla_column_family_tablet_count{cf="mytable2",ks="testing",shard="1"} 1.000000
scylla_column_family_tablet_count{cf="mytable3",ks="testing",shard="1"} 1.000000
scylla_column_family_tablet_count{cf="mytable",ks="testing",shard="2"} 1.000000
scylla_column_family_tablet_count{cf="mytable2",ks="testing",shard="2"} 1.000000
scylla_column_family_tablet_count{cf="mytable3",ks="testing",shard="2"} 1.000000
scylla_column_family_tablet_count{cf="mytable",ks="testing",shard="3"} 1.000000
scylla_column_family_tablet_count{cf="mytable2",ks="testing",shard="3"} 1.000000
scylla_column_family_tablet_count{cf="mytable3",ks="testing",shard="3"} 1.000000
```

Sample output from run without it:
```
# HELP scylla_column_family_tablet_count Tablet count
# TYPE scylla_column_family_tablet_count gauge
scylla_column_family_tablet_count{shard="3"} 3.000000
scylla_column_family_tablet_count{shard="2"} 3.000000
scylla_column_family_tablet_count{shard="1"} 3.000000
scylla_column_family_tablet_count{shard="0"} 3.000000
```

Refs: scylladb#16131

denesb commented:
This can already be obtained from `system.tablets` table. Can grafana extract metrics from tables?

amnonh commented on replica/table.cc:
```diff
@@ -1838,6 +1840,33 @@ void table::update_effective_replication_map(locator::effective_replication_map_
         old_erm->invalidate();
     }
+
+    _stats.tablet_count_on_current_shard = calculate_tablet_count_on_current_shard();
```
That would be executed once and will never be updated

pwrobelse replied:
Okay, so it seems, that I misunderstood Tomek's comment.

> **It could publish table::_tablet_count, which would be set in table::update_effective_replication_map(), which is the place through which all tablet map changes are published.**

Should the calculation be registered via lambda in `table::set_metrics()` then? Would it be okay?

Ref:
https://github.com/scylladb/scylladb/issues/16131#issuecomment-1915564252

tgrabiec replied:
> Should the calculation be registered via lambda in `table::set_metrics()` then? Would it be okay?

No, calculating in update_effective_replication_map() is better, it's more rare than metric scrape period.

pwrobelse replied:
The calculation in `update_effective_replication_map()` has been preserved.

amnonh commented on replica/table.cc:
```diff
@@ -1838,6 +1840,33 @@ void table::update_effective_replication_map(locator::effective_replication_map_
+
+    int64_t new_tablet_count_on_this_shard{0};
+
+    for (auto tablet_id : tablet_map.tablet_ids()) {
```
You can use const reference for the iterator

pwrobelse replied:
Sure, fixed :+1:

tgrabiec replied:
tablet_id is a scalar, access by value is fine.

pwrobelse replied:
Reverted the changes to use auto without reference.

amnonh commented on replica/table.cc:
```diff
@@ -1838,6 +1840,33 @@ void table::update_effective_replication_map(locator::effective_replication_map_
+    _stats.tablet_count_on_current_shard = calculate_tablet_count_on_current_shard();
+}
+
+int64_t table::calculate_tablet_count_on_current_shard() const {
```
Is this calculation done per table?

pwrobelse replied:
In line 1853 tablet_map is retrieved as follows:
`const auto& tablet_map = token_metadata.tablets().get_tablet_map(_schema->id());`

In `./locator/tablets.hh` I can see that `tablet_metadata::get_tablet_map()` takes `table_id` as the parameter.
`const tablet_map& get_tablet_map(table_id id) const;`

The implementation of this member function is as follows:
```
const tablet_map& tablet_metadata::get_tablet_map(table_id id) const {
    try { 
        return _tablets.at(id);
    } catch (const std::out_of_range&) {
        throw_with_backtrace<std::runtime_error>(format("Tablet map not found for table {}", id));
    } 
}
```

So, the answer for this question seems to be yes - the calculation is done per table.

Is it sufficient? Does it need to change?

tgrabiec commented on replica/table.cc:
```diff
@@ -1223,7 +1224,8 @@ void table::set_metrics() {
                         ms::description("Number of tasks waiting to delete sstables from a table"),
-                        [this] { return _sstable_deletion_sem.waiters(); })(cf)(ks)
+                        [this] { return _sstable_deletion_sem.waiters(); })(cf)(ks),
+                ms::make_gauge("tablet_count_on_current_shard", ms::description("Tablet count on current shard"), _stats.tablet_count_on_current_shard)(cf)(ks)
```
The `_on_current_shard` is not necessary, better drop it. All those metrics are for the current shard.

pwrobelse replied:
> The `_on_current_shard` is not necessary, better drop it. All those metrics are for the current shard.

Ok, fixed. Thanks!

tgrabiec commented on replica/table.cc:
```diff
@@ -1838,6 +1840,33 @@ void table::update_effective_replication_map(locator::effective_replication_map_
+        const std::optional<shard_id> shard_id = tablet_map.get_shard(tablet_id, this_host_id);
+        if (!shard_id.has_value()) {
+            continue;
+        }
```
We should revisit this after https://github.com/scylladb/scylladb/issues/16877, and just count non-null storage groups.

tgrabiec reviewed (changes requested):
Just a nit about metric name.

pwrobelse commented:
> Just a nit about metric name.

Fixed. Is it okay now?

Also, is the change in `table::set_metrics()` valid? Currently, the new metric is added under if statement with `_config.enable_metrics_reporting`. Therefore, `.aggregate({seastar::metrics::shard_label})` is not called for the new metric.

tgrabiec commented on replica/table.cc:
```diff
@@ -1223,7 +1224,8 @@ void table::set_metrics() {
                         ms::description("Number of tasks waiting to delete sstables from a table"),
-                        [this] { return _sstable_deletion_sem.waiters(); })(cf)(ks)
+                        [this] { return _sstable_deletion_sem.waiters(); })(cf)(ks),
+                ms::make_gauge("tablet_count", ms::description("Tablet count"), _stats.tablet_count)(cf)(ks)
```
It should be also under enable_node_aggregated_table_metrics, otherwise it's not exported by default.

pwrobelse replied:
> It should be also under enable_node_aggregated_table_metrics, otherwise it's not exported by default.

Updated the patch to contain the mentioned changes :+1:

amnonh commented on replica/table.cc:
```diff
@@ -1272,7 +1274,8 @@ void table::set_metrics() {
                 ms::make_counter("read_latency_count", ms::description("Number of reads"), [this] {return _stats.reads.histogram().count();})(cf)(ks)(node_table_metrics).aggregate({seastar::metrics::shard_label}).set_skip_when_empty(),
-                ms::make_counter("write_latency_count", ms::description("Number of writes"), [this] {return _stats.writes.histogram().count();})(cf)(ks)(node_table_metrics).aggregate({seastar::metrics::shard_label}).set_skip_when_empty()
+                ms::make_counter("write_latency_count", ms::description("Number of writes"), [this] {return _stats.writes.histogram().count();})(cf)(ks)(node_table_metrics).aggregate({seastar::metrics::shard_label}).set_skip_when_empty(),
+                ms::make_gauge("tablet_count", ms::description("Tablet count"), _stats.tablet_count)(cf)(ks)(node_table_metrics).aggregate({seastar::metrics::shard_label})
```
Actually, in our case, we don't want to aggregate by shard but by ks and cf.

We always want per-shard metrics but only show the ks/cf if it's enabled.

pwrobelse replied:
Fixed :+1:

pwrobelse commented:
Still a few adjustments are needed. They will be uploaded tomorrow morning.

scylladb-promoter commented:
#### :green_circle: CI State: SUCCESS

:white_check_mark: - [Build](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/6217/)
:white_check_mark: - [Container Test](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/6217/)
:white_check_mark: - [Unit Tests](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/6217/)
:white_check_mark: - [dtest](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release/7519/)
#### Build Details:
- Duration: 2 hr 6 min
- Builder: spider5.cloudius-systems.com

amnonh commented on replica/table.cc:
```diff
@@ -1272,7 +1274,8 @@ void table::set_metrics() {
                 ms::make_counter("read_latency_count", ms::description("Number of reads"), [this] {return _stats.reads.histogram().count();})(cf)(ks)(node_table_metrics).aggregate({seastar::metrics::shard_label}).set_skip_when_empty(),
-                ms::make_counter("write_latency_count", ms::description("Number of writes"), [this] {return _stats.writes.histogram().count();})(cf)(ks)(node_table_metrics).aggregate({seastar::metrics::shard_label}).set_skip_when_empty()
+                ms::make_counter("write_latency_count", ms::description("Number of writes"), [this] {return _stats.writes.histogram().count();})(cf)(ks)(node_table_metrics).aggregate({seastar::metrics::shard_label}).set_skip_when_empty(),
+                ms::make_gauge("tablet_count", ms::description("Tablet count"), _stats.tablet_count)(cf)(ks)(node_table_metrics).aggregate({column_family_label, keyspace_label})
```
please remove the ```node_table_metric``` label and take it out of the guarding ```_config.enable_node_aggregated_table_metrics```.

you will want to report this metric always and it's not per table (like the label suggest)

pwrobelse replied:
Sure, fixed :+1:

tgrabiec commented on replica/table.cc:
```diff
@@ -1272,7 +1274,8 @@ void table::set_metrics() {
                 ms::make_counter("read_latency_count", ms::description("Number of reads"), [this] {return _stats.reads.histogram().count();})(cf)(ks)(node_table_metrics).aggregate({seastar::metrics::shard_label}).set_skip_when_empty(),
-                ms::make_counter("write_latency_count", ms::description("Number of writes"), [this] {return _stats.writes.histogram().count();})(cf)(ks)(node_table_metrics).aggregate({seastar::metrics::shard_label}).set_skip_when_empty()
+                ms::make_counter("write_latency_count", ms::description("Number of writes"), [this] {return _stats.writes.histogram().count();})(cf)(ks)(node_table_metrics).aggregate({seastar::metrics::shard_label}).set_skip_when_empty(),
+                ms::make_gauge("tablet_count", ms::description("Tablet count"), _stats.tablet_count)(cf)(ks)(node_table_metrics).aggregate({column_family_label, keyspace_label})
```
Now that I think about it, we should export this metric unconditionally, outside enable_node_aggregated_table_metrics flag. This flag exports per-table metrics and could be disabled in some clusters but we want tablet count in all clusters.

tgrabiec commented on replica/table.cc:
```diff
@@ -1831,13 +1834,47 @@ table::table(schema_ptr schema, config config, lw_shared_ptr<const storage_optio
+}
+
+void table::recalculate_tablet_count_stats() {
+    if (!is_internal_keyspace(_schema->ks_name())) {
```
Why this check?

pwrobelse replied:
The outcome of the offline discussion with @amnonh is that the new metric is not exposed for tables from internal keyspace. In `table::set_metrics()` we expose it depending on `is_internal_keyspace`.

I thought, that there was no point in recalculating the value for these tables - currently we iterate over map and query shard_id of a given tablet. My idea was to save that time.

I can revert it if this is not needed.

tgrabiec replied:
We still want the per-node metric to include all tablets, so I think filtering should not be done at this level.

tgrabiec replied:
In the future we will move internal keyspaces to tablets too. Why do we exclude them in metrics @amnonh ?

amnonh replied:
@tgrabiec It's a lot of metrics that are not very interesting; system tables are always local there are no tablets migration between nodes. We don't want to overwhelm the monitoring server with all those metrics.

This is especially important for users with only a few tables, in those situations, the user can choose to enable per-table metrics even if their monitoring server is not very large.

tgrabiec replied:
They are interesting if you want to debug a problem caused by internal tables.

amnonh replied:
@tgrabiec, and then you have the ```system.tablets```. This is why we removed the internal tables from the reporting. It does more harm than good. This is not just for tablets, this is for most per-table metrics.

tgrabiec replied:
What's the problem with `system.tablets`?

pwrobelse replied:
Would it help if we checked `uses_tablets()` instead of `is_internal_keyspace()` in `set_metrics()`? This way the metric would not be exported for tables that do not use tablets.

I have just tested it. The output seems to be the same in comparison with the previous version.

Is it future-proof? I mean, when the internal keyspace is moved to tablets, then the metric will include them.

pwrobelse replied:
@amnonh @tgrabiec - do we have a decision here? Which version should be implemented? Do I understand correctly, that currently this is the last thing that needs to be adjusted in this PR?

tgrabiec replied:
> Would it help if we checked `uses_tablets()` instead of `is_internal_keyspace()` in `set_metrics()`? This way the metric would not be exported for tables that do not use tablets.

Yes, it makes sense.

pwrobelse replied:
> > Would it help if we checked `uses_tablets()` instead of `is_internal_keyspace()` in `set_metrics()`? This way the metric would not be exported for tables that do not use tablets.
> 
> Yes, it makes sense.

Okay - it is already implemented. I uploaded it with the last patch-set.

pwrobelse commented:
Hi @tgrabiec @amnonh, please find the list of updates that the new patch brings as well as the output from `http://localhost:9180/metrics`.

1. As discussed offline with Amnon, the creation of the new metric has been moved and now it is not being reported for tables from internal keyspace.
2. Changed the aggregation to use `{column_family_label, keyspace_label}`.
3. Introduced calculation of tablet count at the end of constructor of table object. Without it, when table was created the metrics reported via `http://localhost:9180/metrics` did not contain the information about the new tablets.

Please find below the output from the web browser. I added three tables to `testing` keyspace: `mytable`, `mytable2` and `mytable3`.

The application was run with:
`sudo ./tools/toolchain/dbuild ./build/dev/scylla --workdir tmp --smp 4 --developer-mode 1 --experimental-features=consistent-topology-changes --experimental-features=tablets`

Results:
```
# HELP scylla_column_family_tablet_count Tablet count
# TYPE scylla_column_family_tablet_count gauge
scylla_column_family_tablet_count{shard="3"} 3.000000
scylla_column_family_tablet_count{shard="2"} 3.000000
scylla_column_family_tablet_count{shard="1"} 3.000000
scylla_column_family_tablet_count{shard="0"} 3.000000
```


Another run with the following command:
`sudo ./tools/toolchain/dbuild ./build/dev/scylla --workdir tmp --smp 4 --developer-mode 1 --experimental-features=consistent-topology-changes --experimental-features=tablets --enable-keyspace-column-family-metrics true`

Results:
```
# HELP scylla_column_family_tablet_count Tablet count
# TYPE scylla_column_family_tablet_count gauge
scylla_column_family_tablet_count{cf="mytable",ks="testing",shard="0"} 1.000000
scylla_column_family_tablet_count{cf="mytable2",ks="testing",shard="0"} 1.000000
scylla_column_family_tablet_count{cf="mytable3",ks="testing",shard="0"} 1.000000
scylla_column_family_tablet_count{cf="mytable",ks="testing",shard="1"} 1.000000
scylla_column_family_tablet_count{cf="mytable2",ks="testing",shard="1"} 1.000000
scylla_column_family_tablet_count{cf="mytable3",ks="testing",shard="1"} 1.000000
scylla_column_family_tablet_count{cf="mytable",ks="testing",shard="2"} 1.000000
scylla_column_family_tablet_count{cf="mytable2",ks="testing",shard="2"} 1.000000
scylla_column_family_tablet_count{cf="mytable3",ks="testing",shard="2"} 1.000000
scylla_column_family_tablet_count{cf="mytable",ks="testing",shard="3"} 1.000000
scylla_column_family_tablet_count{cf="mytable2",ks="testing",shard="3"} 1.000000
scylla_column_family_tablet_count{cf="mytable3",ks="testing",shard="3"} 1.000000
```

scylladb-promoter commented:
#### :red_circle: CI State: FAILURE

:white_check_mark: - [Build](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/6247/)
:white_check_mark: - [Container Test](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/6247/)
:white_check_mark: - [dtest](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release/7548/)
:x: - [Unit Tests](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/6247/)
#### Failed Tests (2/24144):
- [test_recover_stuck_raft_recovery](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/6247/testReport/junit/%28root%29/test_raft_recovery_stuck/Tests___Unit_Tests___test_recover_stuck_raft_recovery) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_recover_stuck_raft_recovery)
- [topology_custom.test_raft_recovery_stuck.debug.2](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/6247/testReport/junit/%28root%29/non-boost%20tests/Tests___Unit_Tests___topology_custom_test_raft_recovery_stuck_debug_2) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+topology_custom.test_raft_recovery_stuck.debug.2)
#### Build Details:
- Duration: 3 hr 17 min
- Builder: i-0c1ee742f705577b1 (m5d.8xlarge)

pwrobelse commented:
> #### 🔴 CI State: FAILURE
> 
> ✅ - [Build](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/6247/) ✅ - [Container Test](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/6247/) ✅ - [dtest](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release/7548/) ❌ - [Unit Tests](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/6247/)
> #### Failed Tests (2/24144):
> 
>     * [test_recover_stuck_raft_recovery](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/6247/testReport/junit/%28root%29/test_raft_recovery_stuck/Tests___Unit_Tests___test_recover_stuck_raft_recovery) [🔍](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_recover_stuck_raft_recovery)
> 
>     * [topology_custom.test_raft_recovery_stuck.debug.2](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/6247/testReport/junit/%28root%29/non-boost%20tests/Tests___Unit_Tests___topology_custom_test_raft_recovery_stuck_debug_2) [🔍](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+topology_custom.test_raft_recovery_stuck.debug.2)
> 
> 
> #### Build Details:
> 
>     * Duration: 3 hr 17 min
> 
>     * Builder: i-0c1ee742f705577b1 (m5d.8xlarge)



The logs from both failed tests contain the information about unavailable host.

**_test_raft_recovery_stuck.test_recover_stuck_raft_recovery_**:
```
cassandra.cluster.NoHostAvailable: 'Unable to complete the operation against any hosts', ConnectionException('Host has been marked down or removed')})
```

**_test_raft_recovery_stuck.debug.2_**
```
ConnectionRefusedError: [Errno 111] Tried connecting to [IP]. Last error: Connection refused
```

pwrobelse commented:
Restarted the test via "Build with Parameters":
https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/6254/

pwrobelse commented:
@amnonh would you take a look at the latest patch-set? Are any additional changes required?

amnonh commented:
We can merge. We may need to break the registration into different scenarios, but it might be an early optimization, so let's start with this PR and see later.

scylladb-promoter commented:
#### :green_circle: CI State: SUCCESS

:white_check_mark: - [Build](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/6254/)
:white_check_mark: - [Container Test](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/6254/)
:white_check_mark: - [dtest](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release/7554/)
:white_check_mark: - [Unit Tests](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/6254/)
#### Build Details:
- Duration: 1 hr 48 min
- Builder: spider2.cloudius-systems.com
